Add option to retain face data

ComFreek opened the pull request:
(Follow-up from #73.)

## Naming comments
I decided to choose `consume_faces` over `collect_faces` to remain consistent in the naming scheme already used for `consume_vertices` and other methods.

## Tests fail!

Strangely enough, running the tests with `nosetests` will fail in `test_parsers.py`/class `TestParsers`. It seems that `consume_faces` doesn't get propagated to the Material objects.
However, running the class `TestParsers` in isolation with `noosetests .\test\test_parser.py:TestParsers` works.

I also tried `cache=False`, but that didn't change anything.

einarf commented on README.md:
```diff
@@ -36,6 +36,7 @@ A more complex example
 * `strict` (Default: `False`) will raise an exception if unsupported features are found in the obj or mtl file
 * `encoding` (Default: `utf-8`) of the obj and mtl file(s)
 * `create_materials` (Default: `False`) will create materials if mtl file is missing or obj file references non-existing materials
+* `consume_faces` (Default: `False`) will consume a face list for every material as well. Beware that it will greatly increase memory usage for large models.
```
I'm not sure if "consume" is an understandable word to use here. If I didn't know the details of the parser I would be confused about this option.

einarf replied:
Personally I liked "collect" much better. "consuming" is just a parser term of processing data and not necessarily related to what we do with the data itself.

ComFreek replied:
Sure, renamed it back.

einarf commented on pywavefront/obj.py:
```diff
@@ -336,6 +352,9 @@ def consume_faces(self):
             triangulate = (len(self.values) - 1) > 3
 
+            if consumed_faces is not None:
+                consumed_faces.append([None] * (len(self.values) - 1))
```
Personally I would collect the face list as triangles only. Then we know each face is 3 elements in the face list. Right now for some models it would be impossible to understand the face list as some ``f`` statements can contain a variable amount of indices for each ``f`` statement (3 or more). Ensure you ``append`` inside the triangulate check as well.

I would also keep it simple and just ``append()`` to ``consumed_faces`` instead of pre-allocate space like this. Python lists over-allocate internally so memory allocations will be less and less frequent as the list increases in size. It makes more sense if you pre-allocate a lot of indices with ``list.extend``.

ComFreek replied:
> Right now for some models it would be impossible to understand the face list as some f statements can contain a variable amount of indices for each f statement (3 or more).

Yes, this was actually intended. I am not sure how much preprocessing this library is supposed to do. Maybe someone (like me) just wants to access the raw face data from a Wavefront file. Especially, I could not find any other Python Wavefront parsing library.
In #73 it sounded a lot that this library is mainly targeted at "preparing interleaved vertex data". I described my use case [in this comment at #73](https://github.com/greenmoss/PyWavefront/issues/73#issue-346060142). I might be lucky that my Wavefront file only has triangle faces anyway, but if it had quad faces, I would have rolled my own library and thus duplicating a lot of raw Wavefront parsing. This seems unnecessary to me this project could distinguish between "raw parsing" and "preprocessing".

> I would also keep it simple and just append() to consumed_faces instead of pre-allocate space like this.
Note that I don't preallocate the overall face list, but I preallocate the list of vertex indices of each and every face, i.e. I preallocate lists of 3 elements most of the time for my Wavefront files.  So no amortization effects apply. But sure, one might call this "premature optimization". I'd argue in favor of keeping it because it is a very trivial line and also not decreasing readability/maintainbility.

einarf commented on pywavefront/obj.py:
```diff
@@ -336,6 +352,9 @@ def consume_faces(self):
             triangulate = (len(self.values) - 1) > 3
 
+            if collected_faces is not None:
+                collected_faces.append([None] * (len(self.values) - 1))
```
```py
1 2 3       # 1 triangle, 3 indices
3 4 5 6     # 2 triangles, 6 indices
7 8 9 10 11 # 3 triangles, 9 indices
```

ComFreek replied:
Thanks for the example! Did you miss my previous comment by any chance ([link](https://github.com/greenmoss/PyWavefront/pull/74#discussion_r207511302))? What do you think?

einarf replied:
Ah yes I missed it.

I think we should at least triangulate when collecting faces so new features can possibly be built on top of that.

einarf commented on test/test_parser.py:
```diff
@@ -12,7 +12,7 @@ def prepend_dir(file):
     def setUp(self):
         # Append current path to locate files
-        meshes = pywavefront.Wavefront(prepend_dir('simple.obj'))
+        meshes = pywavefront.Wavefront(prepend_dir('simple.obj'), collect_faces=True, cache=False)
```
I would make a separate test class for ``collect_faces``. This way we test the parser with and without the new feature.

```py
class TestParserCollectFaces(TestParsers):
    # Override setup
    def setUp(self):
        meshes = pywavefront.Wavefront(prepend_dir('simple.obj'), collect_faces=True)
        self.mesh1 = meshes.mesh_list[0]
        self.mesh2 = meshes.mesh_list[1]

    # ... add your own test methods here ...
```

einarf replied:
The reason you are getting errors is that other tests also extend ``TestParsers`` overriding the ``setUp`` methods, so they sabotage they tests you added in ``TestParsers``. That is why you need to make your own class for testing the feature 😄

ComFreek replied:
Alright! Will do.

ComFreek commented:
@einarf 

1. [x] Separated my tests. I deliberately avoided the inheritance from `TestParsers` because I added a new Wavefront file for testing.
2. [x] Implemented triangulating. Is he order of the triangulated faces OK with you? I applied the same ordering as the vertex emitting, see the code in `obj.py`, line 417.
3. [ ] Documentation for README and `obj.py` (i.e. `consume_faces` docstring) need to be updated. I wanted to wait for your opinion on 2 before I do that.

greenmoss commented:
Are there any metrics on python2 vs python3 support in other projects? What is the "best practice"?

Trivia: while investigating this question, I found https://pythonclock.org

greenmoss commented:
LGTM as well. Ready to merge, pending either a passing test, or a compelling reason to drop Python 2 support.

einarf commented:
I would suggest: Release a new major version: 1.0.0 that is python 3 only, update README and tests.
Maybe also create a 0.4.x branch here just in case the python 2 version need tweaks.

We can simplify a lot of things in the current code base with py3. Having ``yield from`` alone simplifies things so much. The code could need a quick cleanup round as well removing unnecessary stuff.

einarf commented:
I guess the python version info in readme is a bit confusing. Feel free to tweak.

ComFreek commented:
@greenmoss Is there any consensus on how to continue from here?

@einarf I am not sure if rewriting this PR to Python 2 is worthwhile if the project will switch to Python 3 anyway.

einarf commented:
This is up to @greenmoss really.

Option 1:
* Make a 4.x branch here before merging this and release 4.2. Keep the 4.2 branch in case python 2 fixes are needed. (``git checkout -b 0.4.x``, ``git push origin 0.4.x``, ``git checkout master``)
* Merge this PR and release a 1.0.0 release that is python 3 only (also updating readme)

Then we could move ahead and strip away the python 2 stuff. It doesn't create any issues as it is now. py2 is EOL soon and most people have moved to python 3 a long time ago.

Option 2:
* Stay with py2/3 for now

Honestly when I think of it, with the activity level of this project and the fact that it's fairly small I would personally pick option 1. It's a driver to make more things happen in the project.

greenmoss commented:
Option 1 is fine for me. I'll see if I can get a release out this evening.

greenmoss commented:
Release `0.4.2` is out, and branch `python2` is created. Feel free to tear out any Python 2-isms that are in your way.

einarf commented:
@greenmoss  There should be no py2 stuff standing in the way. The main thing is to update circleci to use py3 for tests and re-run tests for this PR.

greenmoss commented:
@einarf interesting I see `.circleci/config.yml` on this branch updated to `python:3.4`, but it's still failing tests. Are there other updates needed as well? Or does it need to be something like `python:3.7`?

einarf commented:
#76 should fix the python 3.4 issue. Rebase this PR after merge and we're good for python 3.4 👍

greenmoss commented:
Thanks @einarf ; merged #76

ComFreek commented:
Finally the tests are passing! @greenmoss This PR is ready to merge 😄

greenmoss commented:
@ComFreek there are two un-checked check boxes in the comments above:

- Documentation for README and obj.py (i.e. consume_faces docstring) need to be updated. I wanted to wait for your opinion on 2 before I do that.
- Added a preliminary "Drop Python 2" commit to see how CircleCI behaves - indeed the tests fail for Python 3.4 due to str/bytes.

Are these still pending, or do you want me to merge as is?
